File: modules/core/app/models/MorphiaFactory.java
package models;

import com.mongodb.*;
import exception.AizouException;
import exception.ErrorCode;
import org.mongodb.morphia.Datastore;
import org.mongodb.morphia.Morphia;
import org.mongodb.morphia.ValidationExtension;
import play.Configuration;
import play.Play;

import java.net.UnknownHostException;
import java.util.*;

/**
 * Morphia生成工厂。
 *
 * @author Zephyre
 */
public class MorphiaFactory {

    private static Hashtable<DBType, Datastore> dsMap = new Hashtable<>();
    private static MorphiaFactory ourInstance;
    private final Morphia morphia;
    private final MongoClient client;

//    private MorphiaFactory(String host, int port) throws TravelPiException {
//        try {
//            client = new MongoClient(host, port);
//        } catch (UnknownHostException e) {
//            throw new TravelPiException(ErrorCode.DATABASE_ERROR, "Invalid database connection.");
//        }
//
//        morphia = new Morphia();
//        new ValidationExtension(morphia);
//    }

    private MorphiaFactory() throws AizouException {
        Configuration config = Play.application().configuration();

        List<ServerAddress> servers = new ArrayList<>();
        Object confEntry = ((Map) config.asMap().get("mongodb")).get("serverList");
        if (confEntry == null) {
            Map<String, Object> tmp = new HashMap<>();
            tmp.put("host", "localhost");
            tmp.put("port", 27017);
            confEntry = Arrays.asList(tmp);
        }
        for (Object entry : (List) confEntry) {
            Map s = (Map) entry;
            servers.add(new ServerAddress(s.get("host").toString(), Integer.parseInt(s.get("port").toString())));
        }

        if (servers.isEmpty())
            throw new AizouException(ErrorCode.DATABASE_ERROR, "Invalid database connection settings.");

//        val user = conf.getString("yunkai.mongo.user")
//        val password = conf.getString("yunkai.mongo.password")
//        val dbName = conf.getString("yunkai.mongo.db")
//        val credential = MongoCredential.createScramSha1Credential(user, dbName, password.toCharArray)

//        credential = MongoCredential.create
        client = new MongoClient(servers);

        morphia = new Morphia();
        new ValidationExtension(morphia);

        morphia.mapPackage("models.geo", true);
        morphia.mapPackage("models.guide", true);
        morphia.mapPackage("models.misc", true);
        morphia.mapPackage("models.plan", true);
        morphia.mapPackage("models.poi", true);
        morphia.mapPackage("models.traffic", true);
    }

    public synchronized static MorphiaFactory getInstance() throws AizouException {
        if (ourInstance == null)
            ourInstance = new MorphiaFactory();

        return ourInstance;
    }

//    public synchronized static MorphiaFactory getInstance(String host, int port) throws TravelPiException {
//        if (ourInstance == null)
//            ourInstance = new MorphiaFactory(host, port);
//
//        return ourInstance;
//    }

    public Morphia getMorphia() {
        return morphia;
    }

    public synchronized Datastore getDatastore(DBType type) throws AizouException {
        // 初始化
        getInstance();

        if (dsMap.contains(type))
            return dsMap.get(type);

        Datastore ds = morphia.createDatastore(client, "k2-dev");
        try {
            ds.ensureIndexes();
            ds.ensureCaps();
            dsMap.put(type, ds);
        } catch (MongoClientException e) {
            throw new AizouException(ErrorCode.DATABASE_ERROR, "Database error.", e);
        }

        return ds;
    }

    public enum DBType {
        GEO,
        POI,
        PLAN,
        PLAN_UGC,
        USER,
        MISC,
        TRAFFIC,
        GUIDE,
        IMAGESTORE,
        TRAVELNOTE
    }
}

File: modules/core/app/database/MorphiaFactory.scala
package database

import com.mongodb.{ MongoClient, MongoClientOptions, MongoCredential, ServerAddress }
import misc.CoreConfig
import org.mongodb.morphia.{ Morphia, ValidationExtension }

import scala.collection.JavaConversions._
import scala.language.postfixOps

/**
 * Created by zephyre on 6/28/15.
 */
object MorphiaFactory {
  lazy val client = {
    val backends = CoreConfig.conf.getConfig("backends.mongo").get
    val services = backends.subKeys.toSeq map (backends.getConfig(_).get)

    val serverAddresses = services map (c => {
      new ServerAddress(c.getString("host").get, c.getInt("port").get)
    })

    val mongoConfig = CoreConfig.conf.getConfig("k2.mongo").get
    val user = mongoConfig.getString("user")
    val password = mongoConfig.getString("password")
    val dbName = mongoConfig.getString("db").get
    val credentialOpt = for {
      u <- user
      p <- password
    } yield {
      MongoCredential.createScramSha1Credential(u, dbName, p.toCharArray)
    }

    val options = new MongoClientOptions.Builder()
      //连接超时
      .connectTimeout(60000)
      //IO超时
      .socketTimeout(10000)
      //与数据库能够建立的最大连接数
      .connectionsPerHost(50)
      //每个连接可以有多少线程排队等待
      .threadsAllowedToBlockForConnectionMultiplier(50)
      .build()

    if (credentialOpt nonEmpty)
      new MongoClient(serverAddresses, Seq(credentialOpt.get), options)
    else
      new MongoClient(serverAddresses, options)
  }

  lazy val morphia = {
    val m = new Morphia()
    new ValidationExtension(m)
    m.mapPackage("models.geo", true)
    m.mapPackage("models.guide", true)
    m.mapPackage("models.misc", true)
    m.mapPackage("models.plan", true)
    m.mapPackage("models.poi", true)
    m.mapPackage("models.traffic", true)
    m
  }

  lazy val datastore = {
    val dbName = CoreConfig.conf.getString("k2.mongo.db").get
    val d = morphia.createDatastore(client, dbName)
    //    d.ensureIndexes()
    d
  }

}
